File: 1.x/localsim-web/models/infra/road/road.js
class AbstractRoad extends Signal{
    constructor(label, src, dst, lanes, lane_width, speed_limit, priority, type_, z_axis, split_nodes = []){
        super();
        this.events = ['destroy', 'change', 'move'];
        this.label = label;
        this._lanes = lanes;
        this._lane_width = lane_width;
        this.width = lanes * lane_width;
        this.speed_limit = speed_limit;
        this._priority = priority;
        this.type_ = type_;
        this.z_axis = z_axis;
        this.segments = new SegmentCollection(src, dst, split_nodes);
        this.src = this.segments.src;
        this.dst = this.segments.dst;
    }

    get_priority(){
        return this._priority;
    }

    set_priority(new_priority){
        if(new_priority != this.get_priority()){
            this._priority = new_priority;
            this.fire('change', new Map().set('value', 'priority'))
        }
    }

    get_lanes(){
        return this._lanes;
    }
    set_lanes(new_lane){
        if(new_lane > 0){
            if(new_lane != this.get_lanes()){
                this._lanes = new_lane;
                this.width = this.get_lanes() * this.get_lane_width();
                this.fire('change', new Map().set('value', 'lanes'));
            }
        }
        else{
            throw 'ValueError';
        }
    }

    get_lane_width(){
        return this._lane_width;
    }

    set_length(d2){
        let segment = this.segments.segments[this.segments.segments.length-1];
        let x1 = segment.src.x;
        let y1 = segment.src.y;
        let x2 = segment.dst.x;
        let y2 = segment.dst.y;
        let d1 = segment.get_length();
        if(this.segments.segments.length > 1){
            for(let i=0; i<this.segments.segments.length-1; i++){
                let s = this.segments.segments[i];
                d2 -= s.get_length();
            }
        }

        let new_x = x1 + ((x2 - x1) * (d2 / d1));
        let new_y = y1 + ((y2 - y1) * (d2 / d1));
        segment.set_length(d2);
        segment.dst.move(new_x, new_y);
    }

    set_lane_width(new_lane_width){
        if(new_lane_width > 0){
            if(new_lane_width != this._lane_width){
                this._lane_width = new_lane_width;
                this.width = this._lanes * this._lane_width;
                this.fire('change', new Map().set('value', 'lane_width'));
            }
        }
        else{
            throw 'ValueError';
        }
    }

    destroy(){
        this.fire('destroy');
    }
}


class Uninterrupted_Road extends AbstractRoad{
    constructor(label, src, dst, lanes, lane_width, speed_limit, priority, type_, z_axis, split_nodes){
        super(label, src, dst, lanes, lane_width, speed_limit, priority, type_, z_axis, split_nodes);

        this.src.connect(this, 'move', this._signal_callback);
        this.dst.connect(this, 'move', this._signal_callback);

        this.length = this.segments.get_length();
        let s_nodes = this.segments.get_split_nodes();

        for(let s_node of s_nodes){
            s_node.connect(this, 'move', this._signal_callback);
        }
    }

    _signal_callback(event, source, kwargs) {
        if(event == 'move'){
            this.length = this.segments.get_length();
            this.fire('move');
        }
    }

    split(segment, x, y){
        let split_node = this.segments.split(segment, x, y);
        split_node.connect(this, 'move', this._signal_callback);
        this.length = this.segments.get_length();
        this.fire('change', new Map().set('action', 'split'));
        return split_node;
    }

    merge(split_node){
        this.segments.merge(split_node);
        split_node.disconnect(this, 'move', this._signal_callback);
        split_node.destroy();
        this.length = this.segments.get_length();
        this.fire('change', new Map().set('action', 'merge'));
    }

    deconstruct(){
        const cls = "localsim.models.infra.road.road.UninterruptedRoad";
        let targs = [this.label, this.src, this.dst, parseInt(this.get_lanes()), this.get_lane_width(), parseFloat(this.speed_limit),
            this.get_priority(), this.type_, this.z_axis];
        const args = targs.concat(this.segments.get_split_nodes());
        const kwargs = {};
        return [cls, args, kwargs]
    }
}


class Interrupted_Road extends AbstractRoad{
    constructor(label, src_road, dst_road, src_list, dst_list, lanes, lane_width, speed_limit, priority, type_, z_axis, split_nodes){
        super(label, src_road.dst, dst_road.src, lanes, lane_width, speed_limit, priority, type_, z_axis, split_nodes);

        //this.w_src_road = new WeakSet().add(src_road);
        //this.w_dst_road = new WeakSet().add(dst_road);
        this.src_road = src_road;
        this.dst_road = dst_road;

        this.src_list = src_list;
        this.dst_list = dst_list;

        this.src_road.connect(this, 'move', this._signal_callback);
        this.dst_road.connect(this, 'move', this._signal_callback);
        this.src_road.connect(this, 'change', this._signal_callback);
        this.dst_road.connect(this, 'change', this._signal_callback);

        this.length = this.segments.get_length();
        let s_nodes = this.segments.get_split_nodes();
        for(let s_node of s_nodes){
            s_node.connect(this, 'move', this._signal_callback);
        }

        src_list.sort();
        dst_list.sort();

        this.in_matrix = new Map();
        this.out_matrix = new Map();

        for(let ctr=0; ctr<src_list.length; ctr++){
            this.in_matrix.set(src_list[ctr], ctr);
            this.out_matrix.set(ctr, dst_list[ctr]);
        }
    }

    _signal_callback(event, source, kwargs){ // source and kwargs
        if(event == 'move'){
            this.length = this.segments.get_length();
            this.fire('move');
        }

        else if(event == 'destroy'){
            this.fire('destroy');
        }

        else if(event == 'change'){ //  and kwargs.get('value') == 'lanes'
            // TODO: Do something
            this.fire('change');
        }
    }

    split(segment, x, y){
        let split_node = this.segments.split(segment, x, y);
        split_node.connect(this, 'move', this._signal_callback);
        this.length = this.segments.get_length();
        this.fire('change', new Map().set('action', 'split')); // action=split
        return split_node;
    }

    merge(split_node){
        this.segments.merge(split_node);
        split_node.disconnect(this, 'move', this._signal_callback);
        split_node.destroy();
        this.length = this.segments.get_length();
        this.fire('change', new Map().set('action', 'merge')); // action=merge
    }

    deconstruct(){
        const cls = "localsim.models.infra.road.road.InterruptedRoad";
        let from = [];
        let to = [];
        for(let [key, value] of this.in_matrix){
            from.push(key);
        }

        for(let [key_, value_] of this.out_matrix){
            to.push(value_);
        }

        let targs = [this.label, this.src_road, this.dst_road, from, to, parseInt(this.get_lanes()),
            this.get_lane_width(), parseFloat(this.speed_limit), this.get_priority(), this.type_, this.z_axis];
        const args = targs.concat(this.segments.get_split_nodes());
        const kwargs = {};
        return [cls, args, kwargs]
    }

    edit_lane_match(src_list, dst_list){
        src_list.sort();
        dst_list.sort();
        this.in_matrix = new Map();
        this.out_matrix = new Map();

        for(let ctr=0; ctr<src_list.length; ctr++){
            this.in_matrix.set(src_list[ctr], ctr);
            this.out_matrix.set(ctr, dst_list[ctr]);
        }
        this.fire('change');
    }
}


class Segment{
    constructor(src, dst){
        this.src = src;
        this.dst = dst;

        this.dx = this.dst.x - this.src.x;
        this.dy = this.dst.y - this.src.y;
        this._length = Math.sqrt((this.dx*this.dx) + (this.dy*this.dy));

        this.src.connect(this, 'move', this._signal_callback);
        this.dst.connect(this, 'move', this._signal_callback);
    }

    _signal_callback(event, source, kwargs){
        if(event == 'move'){
            this.dx = this.dst.x - this.src.x;
            this.dy = this.dst.y - this.src.y;
            this.set_length(Math.sqrt((this.dx*this.dx) + (this.dy*this.dy)));
        }
    }

    split(x, y){
        let split_node = new Node(x, y, 'seg');
        return [new Segment(this.src, split_node), new Segment(split_node, this.dst)];
    }

    merge(segment){
        return new Segment(this.src, segment.dst);
    }

    get_length(){
        return this._length;
    }

    set_length(length){
        this._length = length;
    }
}


class SegmentCollection{
    constructor(src, dst, split_nodes){
        this.src = src;
        this.dst = dst;
        this.segments = [new Segment(this.src, this.dst)];

        for(let ctr=0; ctr<split_nodes.length; ctr++){
            this.split(this.segments[ctr], split_nodes[ctr].x, split_nodes[ctr].y);
        }
    }

    split(segment, x, y){
        let split = segment.split(x, y);
        let split1 = split[0],
            split2 = split[1];

        let index = this.segments.indexOf(segment);

        this.segments[index] = split2;
        this.segments.splice(index, 0, split1);

        return split1.dst;
    }

    get_split_nodes(){
        var ret = [];
        for(let ctr=0; ctr<this.segments.length-1; ctr++){
            ret.push(this.segments[ctr].dst);
        }
        return ret;
    }

    get_split_xy(){
        let ret = [];
        for(let ctr=0; ctr<this.segments.length-1; ctr++){
            ret.push(new Map());

            ret[ctr].set('x', this.segments[ctr].dst.x);
            ret[ctr].set('y', this.segments[ctr].dst.y);
        }
        return ret;
    }

    merge(split_node){
        let split1, split2, split_index;
        let success = false;

        for(let ctr=0; ctr<this.segments.length; ctr++){
            if(this.segments[ctr].dst == split_node){
                split1 = this.segments[ctr];
                split2 = this.segments[ctr + 1];
                split_index = ctr;
                success = true;
                break;
            }
        }

        if(success == false){
            throw 'Merge Error';
        }

        this.segments[split_index] = split1.merge(split2);
        this.segments.splice(this.segments.indexOf(split2), 1);
    }

    get_length(){
        let sum = 0;
        for(let seg of this.segments){
            sum += seg.get_length();
        }

        return sum.toFixed(2);
    }
}

File: 1.x/localsim/models/infra/road/road.py
import weakref
from abc import ABCMeta, abstractmethod
from math import sqrt
from localsim.utils import signal


class AbstractRoad(signal.Signal):
    __metaclass__ = ABCMeta

    events = ['destroy', 'change', 'move']

    def __init__(self, label, src, dst, lanes, lane_width, speed_limit, priority, type_, z_axis, *split_nodes):
        self._lane_width = 0
        self._lanes = 0
        self.width = 0
        self.label = label
        self.lanes = lanes
        self.lane_width = lane_width
        self.speed_limit = speed_limit
        self.segments = SegmentCollection(src, dst, *split_nodes)
        self._priority = priority
        self.type = type_
        self._z_axis = z_axis  # [-4, 5]

    @abstractmethod
    def deconstruct(self):
        pass

    @property
    def src(self):
        return self.segments.src

    @property
    def dst(self):
        return self.segments.dst

    @abstractmethod
    def split(self, segment, x, y):
        pass

    @abstractmethod
    def merge(self, split_node):
        pass

    # Do not delete! Needed for conflict zones
    @property
    def priority(self):
        return self._priority

    # Do not delete! Needed for conflict zones
    @priority.setter
    def priority(self, new_priority):
        if new_priority != self._priority:
            self._priority = new_priority
            self.fire('change', value='priority')

    @property
    def lanes(self):
        return self._lanes

    @lanes.setter
    def lanes(self, new_lane):
        if new_lane > 0:
            if new_lane != self._lanes:
                self._lanes = new_lane
                self.width = self._lanes * self._lane_width
                self.fire('change', value='lanes')
        else:
            raise ValueError

    @property
    def lane_width(self):
        return self._lane_width

    @lane_width.setter
    def lane_width(self, new_lane_width):
        if new_lane_width > 0:
            if new_lane_width != self._lane_width:
                self._lane_width = new_lane_width
                self.width = self._lanes * self._lane_width
                self.fire('change', value='lane_width')
        else:
            raise ValueError

    @property
    def z_axis(self):
        return self._z_axis

    @z_axis.setter
    def z_axis(self, new_z_axis):
        if new_z_axis != self._z_axis:
            self._z_axis = new_z_axis
            self.fire('change', value='z_axis')

    def __repr__(self):
        return '<%s: %s>' % (self.__class__.__name__, self.label)

    def __str__(self):
        return self.label

    def __del__(self):
        self.fire('destroy')


class UninterruptedRoad(AbstractRoad):
    def __init__(self, label, src, dst, lanes, lane_width, speed_limit, priority, type_, z_axis, *split_nodes):
        super(UninterruptedRoad, self).__init__(label, src, dst, lanes, lane_width,
                                                speed_limit, priority, type_, z_axis, *split_nodes)

        self.src.connect('move', self._signal_callback)
        self.dst.connect('move', self._signal_callback)

        self.length = self.segments.length
        for s_node in self.segments.get_split_nodes():
            s_node.connect('move', self._signal_callback)

    def _signal_callback(self, event, source, **kwargs):
        if event == 'move':
            self.length = self.segments.length
            self.fire('move')

    def split(self, segment, x, y):
        split_node = self.segments.split(segment, x, y)
        split_node.connect('move', self._signal_callback)
        self.length = self.segments.length
        self.fire('change', action='split')
        return split_node

    def merge(self, split_node):
        self.segments.merge(split_node)
        split_node.disconnect('move', self._signal_callback)
        self.length = self.segments.length
        self.fire('change', action='merge')

    def deconstruct(self):
        fullpath = '.'.join([self.__class__.__module__, self.__class__.__name__])
        args = [
            self.label, self.src, self.dst, self.lanes, self.lane_width, self.speed_limit, self.priority, self.type,
            self.z_axis
        ] + self.segments.get_split_nodes()
        kwargs = {}
        return [fullpath, args, kwargs]


class InterruptedRoad(AbstractRoad):
    def __init__(self, label, src_road, dst_road, src_list, dst_list,
                 lanes, lane_width, speed_limit, priority, type_, z_axis, *split_nodes, **extras):
        super(InterruptedRoad, self).__init__(label, src_road.dst, dst_road.src, lanes, lane_width,
                                              speed_limit, priority, type_, z_axis, *split_nodes)
        self.w_src_road = weakref.ref(src_road)
        self.w_dst_road = weakref.ref(dst_road)

        self.src_road.connect('move', self._signal_callback)
        self.dst_road.connect('move', self._signal_callback)
        self.src_road.connect('change', self._signal_callback)
        self.dst_road.connect('change', self._signal_callback)
        self.src_road.connect('destroy', self._signal_callback)
        self.dst_road.connect('destroy', self._signal_callback)

        self.length = self.segments.length
        for s_node in self.segments.get_split_nodes():
            s_node.connect('move', self._signal_callback)

        src_list.sort()
        dst_list.sort()

        self.in_matrix = {k: v for k, v in zip(src_list, range(self.lanes))}
        self.out_matrix = {k: v for k, v in zip(range(self.lanes), dst_list)}

    @property
    def src_road(self):
        return self.w_src_road()

    @property
    def dst_road(self):
        return self.w_dst_road()

    def _signal_callback(self, event, source, **kwargs):
        if event == 'move':
            self.length = self.segments.length
            self.fire('move')
        elif event == 'destroy':
            self.fire('destroy')
        elif event == 'change' and kwargs.get('value') == 'lanes':
            if source == self.src_road:
                key_list = self.in_matrix.keys()
                for k in key_list:
                    if k not in range(self.src_road.lanes):
                        self.in_matrix.pop(k)
                while len(self.in_matrix) != len(self.out_matrix):
                    self.out_matrix.popitem()
            elif source == self.dst_road:
                key_list = self.out_matrix.keys()
                for k in key_list:
                    if self.out_matrix[k] not in range(self.dst_road.lanes):
                        self.out_matrix.pop(k)
                while len(self.out_matrix) != len(self.in_matrix):
                    self.in_matrix.popitem()
            self.fire('change')

    def split(self, segment, x, y):
        split_node = self.segments.split(segment, x, y)
        split_node.connect('move', self._signal_callback)
        self.length = self.segments.length
        self.fire('change', action='split')
        return split_node

    def merge(self, split_node):
        self.segments.merge(split_node)
        split_node.disconnect('move', self._signal_callback)
        self.length = self.segments.length
        self.fire('change', action='merge')

    def deconstruct(self):
        fullpath = '.'.join([self.__class__.__module__, self.__class__.__name__])
        args = [
            self.label, self.src_road, self.dst_road, self.in_matrix.keys(), self.out_matrix.values(),
            self.lanes, self.lane_width, self.speed_limit, self.priority, self.type, self.z_axis
        ] + self.segments.get_split_nodes()

        kwargs = {}
        return [fullpath, args, kwargs]

    def edit_lane_match(self, src_list, dst_list):
        src_list.sort()
        dst_list.sort()
        self.in_matrix = {k: v for k, v in zip(src_list, range(self.lanes))}
        self.out_matrix = {k: v for k, v in zip(range(self.lanes), dst_list)}
        self.fire('change')


class Segment(object):
    def __init__(self, src, dst):
        self.src = src
        self.dst = dst

        self.dx = self.dst.x - self.src.x
        self.dy = self.dst.y - self.src.y
        self.length = sqrt(self.dx ** 2 + self.dy ** 2)

        self.src.connect('move', self._signal_callback)
        self.dst.connect('move', self._signal_callback)

    def _signal_callback(self, event, source, **kwargs):
        if event == 'move':
            self.dx = self.dst.x - self.src.x
            self.dy = self.dst.y - self.src.y
            self.length = sqrt(self.dx ** 2 + self.dy ** 2)

    def split(self, x, y):
        cls = self.src.__class__

        split_node = cls(x, y, 'seg')

        return Segment(self.src, split_node), Segment(split_node, self.dst)

    def merge(self, segment):
        return Segment(self.src, segment.dst)

    def __repr__(self):
        return '<%s: %s-%s>' % (self.__class__.__name__, self.src, self.dst)


class SegmentCollection(object):
    def __init__(self, src, dst, *split_nodes):
        self.src = src
        self.dst = dst
        self.segments = [Segment(self.src, self.dst)]

        for index, split_node in enumerate(split_nodes):
            self.split(self[index], split_node.x, split_node.y)

    def __iter__(self):
        return iter(self.segments)

    def __getitem__(self, item):
        return self.segments[item]

    def get_split_nodes(self):
        return [self.segments[i].dst for i in range(len(self.segments)-1)]

    def get_split_xy(self):
        return [(self.segments[i].dst.x, self.segments[i].dst.y) for i in range(len(self.segments)-1)]

    def split(self, segment, x, y):
        split1, split2 = segment.split(x, y)

        index = self.segments.index(segment)

        self.segments[index] = split2
        self.segments.insert(index, split1)

        return split1.dst

    def merge(self, split_node):
        success = False
        for index, segment in enumerate(self.segments):
            if segment.dst == split_node:
                split1 = segment
                split2 = self.segments[index + 1]
                split_index = index
                success = True
                break

        if not success:
            raise ReferenceError

        merged_segment = split1.merge(split2)
        self.segments[split_index] = merged_segment
        self.segments.remove(split2)

    @property
    def length(self):
        return round(sum([seg.length for seg in self.segments]), 2)

    def __len__(self):
        return len(self.segments)

    def __repr__(self):
        return '<%s: %s>' % (self.__class__.__name__, str(self.segments))
